File: src/lib/snapshots.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Memory snapshot metadata as returned by the Rust backend. Mirrors
 * `SnapshotMeta` in `src-tauri/src/memory/snapshots.rs`.
 */
export interface SnapshotMeta {
  id: string;
  label: string;
  created_unix_ms: number;
  size_bytes: number;
  file_count: number;
  roots: string[];
}

/** Result of restoring a snapshot. */
export interface RollbackReport {
  files_restored: number;
  files_skipped: number;
  errors: string[];
}

export async function createSnapshot(
  label: string,
  activeProject?: string | null,
): Promise<SnapshotMeta> {
  return invoke<SnapshotMeta>("create_snapshot", {
    label,
    activeProject: activeProject ?? null,
  });
}

export async function listSnapshots(): Promise<SnapshotMeta[]> {
  return invoke<SnapshotMeta[]>("list_snapshots");
}

export async function rollbackSnapshot(id: string): Promise<RollbackReport> {
  return invoke<RollbackReport>("rollback_snapshot", { id });
}

export async function deleteSnapshot(id: string): Promise<void> {
  return invoke("delete_snapshot", { id });
}

export async function pruneSnapshots(keep: number): Promise<number> {
  return invoke<number>("prune_snapshots", { keep });
}

export function formatBytes(n: number): string {
  if (n < 1024) return `${n} B`;
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(1)} KB`;
  return `${(n / 1024 / 1024).toFixed(2)} MB`;
}

export { timeAgo } from "@/lib/time";

File: src-tauri/src/commands/snapshots.rs
//! Tauri command surface for memory snapshots. All heavy lifting lives in
//! `crate::memory::snapshots`; this file just adapts paths from the frontend
//! and runs the blocking IO inside `spawn_blocking`.

use crate::memory::snapshots::{self, RollbackReport, SnapshotMeta};
use std::path::PathBuf;

/// Capture every memory source into a new gzipped tarball.
#[tauri::command]
pub async fn create_snapshot(
    label: String,
    active_project: Option<String>,
) -> Result<SnapshotMeta, String> {
    tokio::task::spawn_blocking(move || {
        let proj = active_project.map(PathBuf::from);
        snapshots::create(&label, proj.as_deref())
    })
    .await
    .map_err(|e| format!("join: {e}"))?
}

/// List existing snapshots, newest first.
#[tauri::command]
pub async fn list_snapshots() -> Result<Vec<SnapshotMeta>, String> {
    tokio::task::spawn_blocking(snapshots::list)
        .await
        .map_err(|e| format!("join: {e}"))?
}

/// Restore a snapshot; never overwrites files newer than the snapshot itself
/// and never writes outside the original capture roots.
#[tauri::command]
pub async fn rollback_snapshot(id: String) -> Result<RollbackReport, String> {
    tokio::task::spawn_blocking(move || snapshots::rollback(&id))
        .await
        .map_err(|e| format!("join: {e}"))?
}

#[tauri::command]
pub async fn delete_snapshot(id: String) -> Result<(), String> {
    tokio::task::spawn_blocking(move || snapshots::delete(&id))
        .await
        .map_err(|e| format!("join: {e}"))?
}

/// Keep the newest `keep` snapshots; delete the rest. Returns count deleted.
#[tauri::command]
pub async fn prune_snapshots(keep: usize) -> Result<usize, String> {
    tokio::task::spawn_blocking(move || snapshots::prune(keep))
        .await
        .map_err(|e| format!("join: {e}"))?
}
